File: EquityInsight_BE/equityinsight/src/main/java/app/equityinsight/watchlist/WatchlistService.java
package app.equityinsight.watchlist;

import app.equityinsight.exception.StockNotFoundException;
import app.equityinsight.stock.Stock;
import app.equityinsight.stock.StockService;
import app.equityinsight.stock.dto.StockSummaryDto;
import app.equityinsight.exception.WatchlistNotFoundException;
import app.equityinsight.watchlist.dto.CreateWatchlistDto;
import app.equityinsight.watchlist.dto.UpdateWatchlistDto;
import app.equityinsight.watchlist.dto.WatchlistDto;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;

@Service
@Transactional(readOnly = true)
public class WatchlistService {

    private static final Logger log = LoggerFactory.getLogger(WatchlistService.class);

    private final WatchlistRepository watchlistRepository;
    private final WatchlistMapper watchlistMapper;
    private final StockService stockService;

    public WatchlistService(
            WatchlistRepository watchlistRepository,
            WatchlistMapper watchlistMapper,
            StockService stockService
    ) {
        this.watchlistRepository = watchlistRepository;
        this.watchlistMapper = watchlistMapper;
        this.stockService = stockService;
    }

    public List<WatchlistDto> getAllWatchlists() {
        return watchlistRepository.findAll().stream()
                .map(watchlistMapper::toDto)
                .toList();
    }

    public List<StockSummaryDto> getStocksByWatchlistId(Long id) {
        Watchlist watchlist = getWatchlistById(id);

        return watchlist.getStocks().stream()
                .map(watchlistMapper::toStockSummaryDto)
                .toList();
    }

    @Transactional
    public WatchlistDto createWatchlist(CreateWatchlistDto dto) {
        Watchlist watchlist = new Watchlist(dto.title());
        Watchlist savedWatchlist = watchlistRepository.save(watchlist);
        return watchlistMapper.toDto(savedWatchlist);
    }

    @Transactional
    public void deleteWatchlist(Long id) {
        Watchlist watchlist = getWatchlistById(id);

        for (Stock stock : new HashSet<>(watchlist.getStocks())) {
            stock.removeWatchlist(watchlist);
        }

        watchlistRepository.delete(watchlist);
    }

    @Transactional
    public WatchlistDto updateTitle(Long id, UpdateWatchlistDto dto) {
        Watchlist watchlist = getWatchlistById(id);

        watchlist.setTitle(dto.title());
        Watchlist savedWatchlist = watchlistRepository.save(watchlist);
        return watchlistMapper.toDto(savedWatchlist);
    }

    @Transactional
    public StockSummaryDto addStockToWatchlist(Long watchlistId, String rawTicker) {
        Watchlist watchlist = getWatchlistById(watchlistId);
        Stock stock = stockService.findOrCreateEntityByTicker(rawTicker);

        watchlist.addStock(stock);

        return watchlistMapper.toStockSummaryDto(stock);
    }

    @Transactional
    public void removeStockFromWatchlist(Long watchlistId, Long stockId) {
        Watchlist watchlist = getWatchlistById(watchlistId);
        Stock stock = stockService.getStockEntityById(stockId);

        if (!watchlist.getStocks().contains(stock)) {
            log.warn("Cannot remove stock id={} from watchlist id={} because it is not linked", stockId, watchlistId);
            throw new StockNotFoundException(stockId);
        }

        watchlist.removeStock(stock);
    }

    private Watchlist getWatchlistById(Long id) {
        return watchlistRepository.findById(id)
                .orElseThrow(() -> {
                    log.warn("Watchlist lookup failed for id={}", id);
                    return new WatchlistNotFoundException(id);
                });
    }
}

File: EquityInsight/src/app/services/watchlist.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import {
  WatchlistDto,
  CreateWatchlistDto,
  UpdateWatchlistDto
} from '../models/watchlist.models';
import { StockSummaryDto } from '../models/stock.models';

@Injectable({
  providedIn: 'root'
})
export class WatchlistService {
  private http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/watchlists`;

  getAllWatchlists(): Observable<WatchlistDto[]> {
    return this.http.get<WatchlistDto[]>(this.baseUrl);
  }

  createWatchlist(dto: CreateWatchlistDto): Observable<WatchlistDto> {
    return this.http.post<WatchlistDto>(this.baseUrl, dto);
  }

  updateWatchlistTitle(id: number, dto: UpdateWatchlistDto): Observable<WatchlistDto> {
    return this.http.patch<WatchlistDto>(`${this.baseUrl}/${id}`, dto);
  }

  deleteWatchlist(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  getStocksByWatchlistId(watchlistId: number): Observable<StockSummaryDto[]> {
    return this.http.get<StockSummaryDto[]>(`${this.baseUrl}/${watchlistId}/stocks`);
  }

  addStockToWatchlist(watchlistId: number, ticker: string): Observable<StockSummaryDto> {
    return this.http.post<StockSummaryDto>(
      `${this.baseUrl}/${watchlistId}/stocks?ticker=${encodeURIComponent(ticker)}`,
      {}
    );
  }

  removeStockFromWatchlist(watchlistId: number, stockId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${watchlistId}/stocks/${stockId}`);
  }
}
